code/ai/aicode.cpp

```diff
@@ -3379,33 +3379,36 @@ void ai_form_on_wing(object *objp, object *goal_objp)
 	aip->ai_flags.set(AI::AI_Flags::Formation_object);
 
 	aip->goal_objnum = OBJ_INDEX(goal_objp);
+
+	ai_formation_object_recalculate_slotnums(aip->goal_objnum);
+
 	ai_set_goal_maybe_abort_dock(objp, aip);
 	aip->ok_to_target_timestamp = timestamp(DELAY_TARGET_TIME*4);		//	Super extra long time until can target another ship.
 
 }
 
 /**
- * Given an object and an object on whose wing to form, return slot to use.
- *
- * @todo This function is called per object in formation per frame.  Should store slot in ::ai_info struct.
+ * Redistributes slotnums to all ships flying off a given formation object
+ * In the case this is needed because a ship is dying, its objnum is needed to explicitly exclude it
  */
-int ai_formation_object_get_slotnum(int objnum, object *objp)
+void ai_formation_object_recalculate_slotnums(int form_objnum, int dying_objnum)
```

Comment:
Change `dying_objnum` to `exiting_objnum`, since the ship could also be departing or vanishing.